src/openzaak/components/documenten/api/serializers.py

```diff
@@ -491,19 +751,89 @@ def validate(self, attrs):
+            raise serializers.ValidationError(
+                _("Upload of part files is not complete"), code="incomplete-upload"
+            )
+        is_empty = empty_bestandsdelen and not self.instance.inhoud
+        if is_empty and self.instance.bestandsomvang > 0:
+            raise serializers.ValidationError(
+                _("Either file should be upload or the file size = 0"), code="file-size"
+            )
         return valid_attrs
 
     def save(self, **kwargs):
-        self.instance.unlock_document(
+        # merge files and clean bestandsdelen
+
+        # Because it is a large file upload, the document is immediately locked after
+        # creation. This means that attempting to save the merged inhoud causes CMIS
+        # exceptions (because the document is already locked and thus checked out)
+        # If we do not force unlocking, CMIS will complain about the bestandsomvang not
+        # being the same as the actual file size
+        force_unlock = True if settings.CMIS_ENABLED else self.context["force_unlock"]
+        self.instance.canonical.unlock_document(
             doc_uuid=self.context["uuid"],
             lock=self.context["request"].data.get("lock"),
-            force_unlock=self.context["force_unlock"],
+            force_unlock=force_unlock,
         )
-        self.instance.save()
+        self.instance.canonical.save()
+
+        if settings.CMIS_ENABLED:
+            bestandsdelen = BestandsDeel.objects.filter(
+                informatieobject_uuid=self.instance.uuid
+            ).order_by("volgnummer")
+            empty_parts = bestandsdelen.filter(inhoud="")
+            complete_upload = empty_parts.count() == 0
+            empty_bestandsdelen = empty_parts.count() == bestandsdelen.count()
+        else:
+            bestandsdelen = self.instance.canonical.bestandsdelen.order_by("volgnummer")
+            complete_upload = self.instance.canonical.complete_upload
+            empty_bestandsdelen = self.instance.canonical.empty_bestandsdelen
+
+        if empty_bestandsdelen:
+            return self.instance
+
+        if complete_upload:
+            part_files = [p.inhoud.file for p in bestandsdelen]
+            # create the name of target file using the storage backend to the serializer
+            name = create_filename(self.instance.bestandsnaam)
+            file_field = self.instance._meta.get_field("inhoud")
+            rel_path = file_field.generate_filename(self.instance, name)
+            file_name = os.path.basename(rel_path)
+            # merge files
+            file_dir = os.path.join(settings.PRIVATE_MEDIA_ROOT, file_name)
+            target_file = merge_files(part_files, file_dir, file_name)
+            # save full file to the instance FileField
+            with open(target_file) as file_obj:
+                if settings.CMIS_ENABLED:
+                    self.instance.inhoud = File(file_obj, name=file_name)
+                    self.instance.save()
```

Comment:
I do think the file gets loaded into memory, though I think it behaves the same way as the non-CMIS pattern, so I'll just remove the if/else statement

Reply:
I've looked up the Django source code for filefield, which delegates to the storage save method, ultimately leading to this: https://github.com/django/django/blob/stable/3.2.x/django/core/files/storage.py#L283

Looks like the file is correctly processed in chunks, rather than loaded into memory all at once, so we should be good.